src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/model/TransformMetadata.kt

```diff
@@ -0,0 +1,215 @@
+        if (params.paramAsBoolean(WITH_TYPE, true)) builder.endObject()
+        return builder.endObject()
+    }
+
+    override fun writeTo(out: StreamOutput) {
+        out.writeString(id)
+        out.writeLong(seqNo)
+        out.writeLong(primaryTerm)
+        out.writeString(transformId)
+        out.writeBoolean(afterKey != null)
+        afterKey?.let { out.writeMap(it) }
+        out.writeInstant(lastUpdatedAt)
+        out.writeEnum(status)
+        out.writeOptionalString(failureReason)
+        stats.writeTo(out)
+    }
+
+    companion object {
+        const val NO_ID = ""
+        const val TRANSFORM_METADATA_TYPE = "transform_metadata"
+        const val TRANSFORM_ID_FIELD = "transform_id"
+        const val AFTER_KEY_FIELD = "after_key"
+        const val LAST_UPDATED_AT_FIELD = "last_updated_at"
+        const val STATUS_FIELD = "status"
+        const val STATS_FIELD = "stats"
+        const val FAILURE_REASON = "failure_reason"
+
+        @Suppress("ComplexMethod", "LongMethod")
+        @JvmStatic
+        @Throws(IOException::class)
+        fun parse(
+            xcp: XContentParser,
+            id: String = NO_ID,
+            seqNo: Long = SequenceNumbers.UNASSIGNED_SEQ_NO,
+            primaryTerm: Long = SequenceNumbers.UNASSIGNED_PRIMARY_TERM
+        ): TransformMetadata {
+            var transformId: String? = null
+            var afterkey: Map<String, Any>? = null
+            var lastUpdatedAt: Instant? = null
+            var status: Status? = null
+            var failureReason: String? = null
+            var stats: TransformStats? = null
+
+            ensureExpectedToken(XContentParser.Token.START_OBJECT, xcp.currentToken(), xcp)
+            while (xcp.nextToken() != XContentParser.Token.END_OBJECT) {
+                val fieldName = xcp.currentName()
+                xcp.nextToken()
+
+                when (fieldName) {
+                    TRANSFORM_ID_FIELD -> transformId = xcp.text()
+                    AFTER_KEY_FIELD -> afterkey = xcp.map()
+                    LAST_UPDATED_AT_FIELD -> lastUpdatedAt = xcp.instant()
+                    STATUS_FIELD -> status = Status.valueOf(xcp.text().toUpperCase(Locale.ROOT))
+                    FAILURE_REASON -> failureReason = xcp.textOrNull()
+                    STATS_FIELD -> stats = TransformStats.parse(xcp)
+                }
+            }
+
+            return TransformMetadata(
+                id,
+                seqNo,
+                primaryTerm,
+                transformId = requireNotNull(transformId) { "TransformId must not be null" },
+                afterKey = afterkey,
+                lastUpdatedAt = requireNotNull(lastUpdatedAt) { "Last updated time must not be null" },
+                status = requireNotNull(status) { "Status must not be null" },
+                failureReason = failureReason,
+                stats = requireNotNull(stats) { "Stats must not be null" }
+            )
+        }
+    }
+}
+
+data class TransformStats(
```

Comment:
nit: Could probably split this out to own file

Reply:
Also is this exact copy of the RollupStats by any chance? If so perhaps we can make something like CompositeStats data class that these can extend or something?

Reply:
Sure, will move it. Its not an exact copy, but has the same intent. RollupStats named the field `rollupsIndexed` instead of `documentsIndexed`.
I was wondering about it too, if its okay to rename the field in rollup document (not sure if any issues with changing schema for existing rollup jobs) then I can have have a common data class for both transforms and rollups